types/wlr_input_inhibitor.c

```diff
@@ -0,0 +1,133 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include "wlr/types/wlr_input_inhibitor.h"
+#include "wlr-input-inhibitor-unstable-v1-protocol.h"
+
+static const struct zwlr_input_inhibit_manager_v1_interface inhibit_manager_implementation;
+
+static struct wlr_input_inhibit_manager *input_inhibit_manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_input_inhibit_manager_v1_interface,
+			&inhibit_manager_implementation));
+	return wl_resource_get_user_data(resource);
+}
+
+static void input_inhibitor_destroy(struct wl_client *client,
+			struct wl_resource *resource) {
+	struct wlr_input_inhibit_manager *manager =
+		input_inhibit_manager_from_resource(resource);
+	manager->active_client = NULL;
+	manager->active_inhibitor = NULL;
+	wl_resource_destroy(resource);
+	wl_signal_emit(&manager->events.deactivate, manager);
+}
+
+static struct zwlr_input_inhibitor_v1_interface input_inhibitor_implementation = {
+	.destroy = input_inhibitor_destroy,
+};
+
+static void inhibit_manager_get_inhibitor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id) {
+	struct wlr_input_inhibit_manager *manager =
+		input_inhibit_manager_from_resource(resource);
+	if (manager->active_client || manager->active_inhibitor) {
+		wl_resource_post_error(resource,
+			ZWLR_INPUT_INHIBIT_MANAGER_V1_ERROR_ALREADY_INHIBITED,
+			"this compositor already has input inhibited");
+		return;
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(client,
+			&zwlr_input_inhibitor_v1_interface,
+			wl_resource_get_version(resource), id);
+	if (!wl_resource) {
+		wl_client_post_no_memory(client);
+	}
+	wl_resource_set_implementation(wl_resource, &input_inhibitor_implementation,
+			manager, NULL);
```

Comment:
What happens if the client crashes or disconnects without sending a `destroy` request?

I'd move the `input_inhibitor_destroy` logic into the `wl_resource` destroy signal handler.

Reply:
>What happens if the client crashes or disconnects without sending a destroy request?

https://github.com/swaywm/wlroots/pull/807/files/0e613bed5cf0383450586de9b9219fbf639e8db7#diff-73e9f66aeebd8121768ea744a5bff64fR61